dan_layer/consensus/src/hotstuff/on_receive_vote.rs

```diff
@@ -1,109 +1,142 @@
-        //     // TODO: This should go in a some component that controls message flows and events
-        //     let epoch = self.epoch_manager.current_epoch().await?;
-        //     let committee = self.epoch_manager.get_committee(epoch, shard_id).await?;
-        //     if committee.is_empty() {
-        //         return Err(HotStuffError::NoCommitteeForShard { shard: shard_id, epoch });
-        //     }
-        //     if self.is_leader(payload_id, shard_id, &committee)? {
-        //         self.leader_on_propose(shard_id, payload_id).await?;
-        //     }
-        // }
-        // Ok(())
+
+        let addr = self.epoch_manager.get_our_validator_addr(message.epoch).await?;
+        let committee = self.epoch_manager.get_local_committee(message.epoch).await?;
+        if !committee.contains(&addr) {
+            return Err(HotStuffError::ReceivedMessageFromNonCommitteeMember {
+                epoch: message.epoch,
+                sender: from.to_string(),
+                context: "OnVoteReceived".to_string(),
+            });
+        }
+
+        if !self.leader_strategy.is_leader(&addr, &committee, &message.block_id, 0) {
+            return Err(HotStuffError::NotTheLeader {
+                details: format!("Not this leader for block {}, vote sent by {}", message.block_id, addr),
+            });
+        }
+
+        let local_commmittee_shard = self.epoch_manager.get_local_committee_shard(message.epoch).await?;
+
+        // Get the sender shard, and check that they are in the local committee
+        let sender_shard = self
+            .epoch_manager
+            .get_validator_shard(message.epoch, from.clone())
+            .await?;
+        if !local_commmittee_shard.includes_shard(&sender_shard) {
+            return Err(HotStuffError::ReceivedMessageFromNonCommitteeMember {
+                epoch: message.epoch,
+                sender: from.to_string(),
+                context: "OnVoteReceived".to_string(),
+            });
+        }
+
+        let mut tx = self.store.create_write_tx()?;
+        let block = Block::get(tx.deref_mut(), &message.block_id)?;
+        if block.proposed_by() != local_commmittee_shard.our_shard_id() {
+            return Err(HotStuffError::NotTheLeader {
+                details: format!(
+                    "Block {} was not proposed by this validator {}",
+                    message.block_id,
+                    local_commmittee_shard.our_shard_id()
+                ),
+            });
+        }
+
+        Vote {
+            epoch: message.epoch,
+            block_id: message.block_id,
+            decision: message.decision,
+            sender: sender_shard,
+            signature: message.signature,
+            merkle_proof: message.merkle_proof,
+        }
+        .save(&mut tx)?;
+
+        let count = Vote::count_for_block(tx.deref_mut(), &message.block_id)?;
+
+        if count < local_commmittee_shard.quorum_threshold() {
```

Comment:
This handles the case where you have less than the required votes, but you also need a check for whether you have constructed a QC already and distributed it. 

In other words, when a vote comes in over the threshold, what happens. 
For example, in a 4 member committee, the 3rd vote comes in, we continue the logic below. When the 4th vote comes in, we will repeat this process, when (I believe) we shouldn't

Reply:
ah yup you are right - good catch